### PR TITLE
Database-based Argument Setter

### PR DESCRIPTION
**Pull request by ailegion**
added: Database-based Argument Setter support

JIRA: https://issues.cask.co/browse/PLUGIN-173

### CODE REVIEW

File: database-commons/src/main/java/io/cdap/plugin/db/batch/action/AbstractDBArgumentSetter.java [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+  public void run(ActionContext context) throws Exception {
+    Class<? extends Driver> driverClass = context.loadPluginClass(JDBC_PLUGIN_ID);
+    FailureCollector failureCollector = context.getFailureCollector();
+    SettableArguments settableArguments = context.getArguments();
+    processArguments(driverClass, failureCollector, settableArguments);
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer)
+      throws IllegalArgumentException {
```
Comment (CuriousVini) [inline database-commons/src/main/java/io/cdap/plugin/db/batch/action/AbstractDBArgumentSetter.java]: nit: misallignment

File: database-commons/src/main/java/io/cdap/plugin/db/batch/action/AbstractDBArgumentSetter.java [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    } catch (SQLException e) {
+      collector.addFailure("SQL error while executing query: " + e.getMessage(), null)
+          .withStacktrace(e.getStackTrace());
+    } catch (IllegalAccessException | InstantiationException e) {
+      collector.addFailure("Unable to instantiate JDBC driver: " + e.getMessage(), null)
+          .withStacktrace(e.getStackTrace());
+    } catch (Exception e) {
+      collector.addFailure(e.getMessage(), null).withStacktrace(e.getStackTrace());
+    }
+    config.validate(collector);
```
Comment (CuriousVini) [inline database-commons/src/main/java/io/cdap/plugin/db/batch/action/AbstractDBArgumentSetter.java]: why do we need to call `config.validate(collector)` 2 times?

File: database-commons/src/main/java/io/cdap/plugin/db/batch/action/ArgumentSetterConfig.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+   * Validates config input fields.
+   *
+   * @param collector context failure collector {@link FailureCollector}
+   */
+  public void validate(FailureCollector collector) {
+    if (!containsMacro(CONNECTION_STRING) && Strings.isNullOrEmpty(this.connectionString)) {
+      collector.addFailure("Invalid connection string", "Connection string cannot be empty");
+    }
+    if (!containsMacro(ConnectionConfig.USER) && Strings.isNullOrEmpty(this.user)) {
+      collector.addFailure("Invalid username", "Username cannot be empty");
```
Comment (CuriousVini) [inline database-commons/src/main/java/io/cdap/plugin/db/batch/action/ArgumentSetterConfig.java]: please add period at the end of every failure message and its cause

File: database-commons/src/main/java/io/cdap/plugin/db/batch/action/ArgumentSetterConfig.java [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+    }
+    if (!containsMacro(ARGUMENT_SELECTION_CONDITIONS) && Strings
+        .isNullOrEmpty(this.getArgumentSelectionConditions())) {
+      collector
+          .addFailure("Invalid conditions", "Filter conditions must be specified");
+    }
+
+    collector.getOrThrowException();
+  }
+
```
Comment (CuriousVini) [inline database-commons/src/main/java/io/cdap/plugin/db/batch/action/ArgumentSetterConfig.java]: nit: extra line

File: database-commons/src/main/java/io/cdap/plugin/db/batch/action/AbstractDBArgumentSetter.java [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+      if (!hasRecord) {
+        failureCollector.addFailure("Ne record found",
+            "No data is returned for the argument selection conditions");
+      }
+      if (settableArguments != null) {
+        setArguments(resultSet, failureCollector, settableArguments);
+      }
+      if (resultSet.next()) {
+        failureCollector
+            .addFailure("More than one record found",
```
Comment (CuriousVini) [inline database-commons/src/main/java/io/cdap/plugin/db/batch/action/AbstractDBArgumentSetter.java]: `More than one records found.`

File: database-commons/src/main/java/io/cdap/plugin/db/batch/action/AbstractDBArgumentSetter.java [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+                "The argument selection conditions return multiple rows");
```
Comment (CuriousVini) [inline database-commons/src/main/java/io/cdap/plugin/db/batch/action/AbstractDBArgumentSetter.java]: what is the expected corrective action that needs to be taken from user?

File: database-commons/src/main/java/io/cdap/plugin/db/batch/action/AbstractDBArgumentSetter.java [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+    Properties connectionProperties = new Properties();
+    connectionProperties.putAll(config.getConnectionArguments());
+    try {
+      Connection connection = DriverManager
+          .getConnection(config.getConnectionString(), connectionProperties);
+      Statement statement = connection.createStatement();
+      ResultSet resultSet = statement.executeQuery(config.getQuery());
+      boolean hasRecord = resultSet.next();
+      if (!hasRecord) {
+        failureCollector.addFailure("Ne record found",
```
Comment (CuriousVini) [inline database-commons/src/main/java/io/cdap/plugin/db/batch/action/AbstractDBArgumentSetter.java]: Typo: `No records found.`
Please add period to every failure message and corresponding corrective action

File: database-commons/src/main/java/io/cdap/plugin/db/batch/action/AbstractDBArgumentSetter.java [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+            "No data is returned for the argument selection conditions");
```
Comment (CuriousVini) [inline database-commons/src/main/java/io/cdap/plugin/db/batch/action/AbstractDBArgumentSetter.java]: If there are no records should we return from here and not do any further processing?

File: database-commons/src/main/java/io/cdap/plugin/db/batch/action/AbstractDBArgumentSetter.java [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+    driverCleanup = DBUtils.ensureJDBCDriverIsAvailable(driverClass, config.getConnectionString(),
+        config.jdbcPluginName);
+    Properties connectionProperties = new Properties();
+    connectionProperties.putAll(config.getConnectionArguments());
+    try {
+      Connection connection = DriverManager
+          .getConnection(config.getConnectionString(), connectionProperties);
+      Statement statement = connection.createStatement();
+      ResultSet resultSet = statement.executeQuery(config.getQuery());
+      boolean hasRecord = resultSet.next();
```
Comment (CuriousVini) [inline database-commons/src/main/java/io/cdap/plugin/db/batch/action/AbstractDBArgumentSetter.java:104]: does resultset has - `hasNext()` method? we should use that method.

**Review by CuriousVini — COMMENTED** [review]
A few comments.

File: database-commons/src/main/java/io/cdap/plugin/db/batch/action/AbstractDBArgumentSetter.java (hunk 1 above)
Comment (ailegion) [reply database-commons/src/main/java/io/cdap/plugin/db/batch/action/AbstractDBArgumentSetter.java]: @CuriousVini fixed

File: database-commons/src/main/java/io/cdap/plugin/db/batch/action/AbstractDBArgumentSetter.java (hunk 2 above)
Comment (ailegion) [reply database-commons/src/main/java/io/cdap/plugin/db/batch/action/AbstractDBArgumentSetter.java]: @CuriousVini removed redundant call

File: database-commons/src/main/java/io/cdap/plugin/db/batch/action/ArgumentSetterConfig.java (hunk 3 above)
Comment (ailegion) [reply database-commons/src/main/java/io/cdap/plugin/db/batch/action/ArgumentSetterConfig.java]: @CuriousVini updated all messages

File: database-commons/src/main/java/io/cdap/plugin/db/batch/action/ArgumentSetterConfig.java (hunk 4 above)
Comment (ailegion) [reply database-commons/src/main/java/io/cdap/plugin/db/batch/action/ArgumentSetterConfig.java]: @CuriousVini fixed

File: database-commons/src/main/java/io/cdap/plugin/db/batch/action/AbstractDBArgumentSetter.java (hunk 8 above)
Comment (ailegion) [reply database-commons/src/main/java/io/cdap/plugin/db/batch/action/AbstractDBArgumentSetter.java]: @CuriousVini fixed to return and not do further processing.

File: database-commons/src/main/java/io/cdap/plugin/db/batch/action/AbstractDBArgumentSetter.java (hunk 9 above)
Comment (ailegion) [reply database-commons/src/main/java/io/cdap/plugin/db/batch/action/AbstractDBArgumentSetter.java:104]: @CuriousVini resulset does not have hasNext() instead it has next() which moves the cursor to next record and returns true/false based on whether record is available or not.

File: database-commons/src/main/java/io/cdap/plugin/db/batch/action/AbstractDBArgumentSetter.java (hunk 7 above)
Comment (ailegion) [reply database-commons/src/main/java/io/cdap/plugin/db/batch/action/AbstractDBArgumentSetter.java]: @CuriousVini updated messages to have corrective action.

File: database-commons/src/main/java/io/cdap/plugin/db/batch/action/AbstractDBArgumentSetter.java (hunk 6 above)
Comment (ailegion) [reply database-commons/src/main/java/io/cdap/plugin/db/batch/action/AbstractDBArgumentSetter.java]: @CuriousVini updated

File: database-commons/src/main/java/io/cdap/plugin/db/batch/action/AbstractDBArgumentSetter.java (hunk 5 above)
Comment (ailegion) [reply database-commons/src/main/java/io/cdap/plugin/db/batch/action/AbstractDBArgumentSetter.java]: @CuriousVini fixed

File: database-commons/src/main/java/io/cdap/plugin/db/batch/action/AbstractDBArgumentSetter.java [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+      if (settableArguments != null) {
+        setArguments(resultSet, failureCollector, settableArguments);
+      }
+      if (resultSet.next()) {
+        failureCollector
+          .addFailure("More than one records found.",
+                      "The argument selection conditions must match only one record.");
+      }
+    } catch (SQLException e) {
+      throw new SQLException(e.getMessage(), e.getSQLState(), e.getErrorCode());
```
Comment (CuriousVini) [inline database-commons/src/main/java/io/cdap/plugin/db/batch/action/AbstractDBArgumentSetter.java]: why do we catch the sqlexception and throw the same exception? It should throw same `SQLException`

File: database-commons/src/main/java/io/cdap/plugin/db/batch/action/ArgumentSetterConfig.java [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+      collector.addFailure("Invalid connection string.", "Connection string cannot be empty.");
+    }
+    if (!containsMacro(ConnectionConfig.USER) && Strings.isNullOrEmpty(this.user)) {
+      collector.addFailure("Invalid username.", "Username cannot be empty.");
+    }
+    if (!containsMacro(ConnectionConfig.PASSWORD) && Strings.isNullOrEmpty(this.password)) {
+      collector.addFailure("Invalid password.", "Password cannot be empty.");
+    }
+    if (!containsMacro(DATABASE_NAME) && Strings.isNullOrEmpty(this.getDatabaseName())) {
+      collector.addFailure("Invalid database.", "Invalid database is specified.");
```
Comment (CuriousVini) [inline database-commons/src/main/java/io/cdap/plugin/db/batch/action/ArgumentSetterConfig.java]: `Invalid database is specified.` -> `Invalid database must be specified.`

File: database-commons/src/main/java/io/cdap/plugin/db/batch/action/ArgumentSetterConfig.java [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+    }
+    if (!containsMacro(TABLE_NAME) && Strings.isNullOrEmpty(this.getTableName())) {
+      collector.addFailure("Invalid table.", "Invalid table is specified.");
```
Comment (CuriousVini) [inline database-commons/src/main/java/io/cdap/plugin/db/batch/action/ArgumentSetterConfig.java]: same here

**Review by CuriousVini — APPROVED** [review]
a few comments. Fix and squash the commits

File: database-commons/src/main/java/io/cdap/plugin/db/batch/action/AbstractDBArgumentSetter.java (hunk 10 above)
Comment (ailegion) [reply database-commons/src/main/java/io/cdap/plugin/db/batch/action/AbstractDBArgumentSetter.java]: > a few comments. Fix and squash the commits

@CuriousVini done.

**Review by CuriousVini — APPROVED** [review]

**Pull request merged**